pkg/metrics/events.go

```diff
@@ -0,0 +1,159 @@
+package metrics
+
+// This file contains wrapper functions that will map a Prometheus metric names to its
+// label field, help field and an associated Event.
+
+// MetricType will represent a Prometheus metric.
+// A variable of type MetricType will hold the name of the Prometheus metric as reported.
+type MetricType string
+
+const (
+	SampleCountType MetricType = "sample_count"
+
+	ActionSetBackupCreatedType   MetricType = "actionset_backup_created_count"
+	ActionSetBackupCompletedType MetricType = "actionset_backup_completed_count"
+
+	ActionSetRestoreCreatedType   MetricType = "actionset_restore_created_count"
+	ActionSetRestoreCompletedType MetricType = "actionset_restore_completed_count"
+
+	ActionSetTotalCreatedType   MetricType = "actionset_total_created_count"
+	ActionSetTotalCompletedType MetricType = "actionset_total_completed_count"
```

Comment:
Done.

Reply:
Small note: `…_failures_total` doesn't stand on its own if the intention is to calculate failure rate. A separate completed count would need to be maintained, wouldn't it?

It appears that the way the original `…_completed_count` was implemented would still accomplish the same goal since it includes a `state` that indicates whether it failed or not. This should allow the rate to be calculated as `rate(…_completed_count{state="failure"}[5m]) / rate(…_completed_count[5m])`.

Reply:
I assume `_created_count` is what will be used with `_failures_total` to calculate failure rates. The `state` label works too, but then why have a separate `created` metric? I interpret _completed_ as _succeeded_.

Reply:
Yeah, I'm not sure if/what value the separate `created` metric provides.

But it did seem weird to me to calculate a rate based upon created vs. completed + failed (especially since there's a strange temporal dissonance there…). It seemed more logical to me to compare the number of completed vs. completed + failed.

Reply:
Is completed + failed the same as "created", which is the same as "total"? In my mind, they are. As a start, I want to focus on the error rates i.e., `rate(actionset_failures_total[5m]) / rate(actionset_total[5m])`. It's more commonly used than non-error rates from a system alert perspective.  The `rate(actionset_total)` is also great for detecting saturation and spikes.

At the code level, ideally, there are two places that call `Inc()`; one for `requests_total` when the actionset is initiated, and one for `failures_total` inside the highest level `if err != nil`.

How would the `state` label work on a counter metric, as the actionset goes through different phases throughout its lifecycle? Will there be one data point for each state?